File: scripts/optimize-public-images.mjs
#!/usr/bin/env node
import { readdir, rename, stat, unlink } from "node:fs/promises";
import { tmpdir } from "node:os";
import path from "node:path";

import sharp from "sharp";

const publicFotosDir = path.resolve("public/fotos");
const supportedExtensions = new Set([".jpg", ".jpeg", ".png", ".webp"]);
const maxDimension = 1600;
const jpegQuality = 78;
const webpQuality = 80;
const minSavingsBytes = 2048;

async function pathExists(filePath) {
  try {
    await stat(filePath);
    return true;
  } catch (error) {
    if (error?.code === "ENOENT") {
      return false;
    }
    throw error;
  }
}

async function collectImages(directory) {
  if (!(await pathExists(directory))) {
    return [];
  }

  const entries = await readdir(directory, { withFileTypes: true });
  const images = [];

  await Promise.all(
    entries.map(async (entry) => {
      const entryPath = path.join(directory, entry.name);

      if (entry.isDirectory()) {
        images.push(...(await collectImages(entryPath)));
        return;
      }

      if (entry.isFile() && supportedExtensions.has(path.extname(entry.name).toLowerCase())) {
        images.push(entryPath);
      }
    }),
  );

  return images;
}

function optimizerForExtension(image, extension) {
  const resized = image.rotate().resize({
    width: maxDimension,
    height: maxDimension,
    fit: "inside",
    withoutEnlargement: true,
  });

  if (extension === ".jpg" || extension === ".jpeg") {
    return resized.jpeg({ quality: jpegQuality, mozjpeg: true, progressive: true });
  }

  if (extension === ".webp") {
    return resized.webp({ quality: webpQuality });
  }

  if (extension === ".png") {
    return resized.png({ compressionLevel: 9, palette: true });
  }

  return resized;
}

async function optimizeFile(filePath) {
  const before = (await stat(filePath)).size;
  const extension = path.extname(filePath).toLowerCase();
  const temporaryPath = path.join(
    tmpdir(),
    `fures-image-${process.pid}-${Date.now()}-${Math.random().toString(16).slice(2)}${extension}`,
  );

  try {
    await optimizerForExtension(sharp(filePath, { failOn: "none" }), extension).toFile(temporaryPath);
    const after = (await stat(temporaryPath)).size;

    if (after + minSavingsBytes < before) {
      await rename(temporaryPath, filePath);
      return { before, after, changed: true };
    }

    await unlink(temporaryPath);
    return { before, after: before, changed: false };
  } catch (error) {
    await unlink(temporaryPath).catch(() => {});
    console.warn(`[optimize-images] skipped ${filePath}: ${error.message}`);
    return { before, after: before, changed: false };
  }
}

async function main() {
  if (!(await pathExists(publicFotosDir))) {
    console.log(`[optimize-images] ${publicFotosDir} not found; nothing to optimize.`);
    return;
  }

  const images = await collectImages(publicFotosDir);
  let beforeTotal = 0;
  let afterTotal = 0;
  let changed = 0;

  for (const imagePath of images) {
    const result = await optimizeFile(imagePath);
    beforeTotal += result.before;
    afterTotal += result.after;
    if (result.changed) {
      changed += 1;
    }
  }

  const toMb = (bytes) => (bytes / 1048576).toFixed(2);
  console.log(
    `[optimize-images] processed=${images.length} changed=${changed} ` +
      `before=${toMb(beforeTotal)}MB after=${toMb(afterTotal)}MB saved=${toMb(beforeTotal - afterTotal)}MB`,
  );
}

await main();

File: scripts/optimize_public_images.py
#!/usr/bin/env python3
"""Compress generated deploy images without changing their public URLs."""
from __future__ import annotations

import os
from pathlib import Path
from tempfile import NamedTemporaryFile

try:
    from PIL import Image, ImageOps
except ImportError as exc:  # pragma: no cover - environment guard for local installs
    print(f"[optimize-images] Pillow is not installed; skipping image optimization: {exc}")
    raise SystemExit(0)

PUBLIC_FOTOS_DIR = Path("public/fotos")
SUPPORTED_EXTENSIONS = {".jpg", ".jpeg", ".png", ".webp"}
MAX_DIMENSION = 1600
JPEG_QUALITY = 78
WEBP_QUALITY = 80
MIN_SAVINGS_BYTES = 2048


def resize_if_needed(image: Image.Image) -> Image.Image:
    width, height = image.size
    largest_side = max(width, height)

    if largest_side <= MAX_DIMENSION:
        return image

    scale = MAX_DIMENSION / largest_side
    resized_size = (max(1, round(width * scale)), max(1, round(height * scale)))
    return image.resize(resized_size, Image.Resampling.LANCZOS)


def save_optimized(image: Image.Image, suffix: str, destination: Path) -> None:
    suffix = suffix.lower()

    if suffix in {".jpg", ".jpeg"}:
        if image.mode not in {"RGB", "L"}:
            image = image.convert("RGB")
        image.save(destination, format="JPEG", quality=JPEG_QUALITY, optimize=True, progressive=True)
        return

    if suffix == ".webp":
        image.save(destination, format="WEBP", quality=WEBP_QUALITY, method=6)
        return

    if suffix == ".png":
        image.save(destination, format="PNG", optimize=True)
        return

    raise ValueError(f"Unsupported image type: {suffix}")


def optimize_file(path: Path) -> tuple[int, int]:
    original_size = path.stat().st_size

    try:
        with Image.open(path) as opened:
            image = ImageOps.exif_transpose(opened)
            image = resize_if_needed(image)

            with NamedTemporaryFile(delete=False, suffix=path.suffix) as temporary_file:
                temp_path = Path(temporary_file.name)

            try:
                save_optimized(image, path.suffix, temp_path)
                optimized_size = temp_path.stat().st_size

                if optimized_size + MIN_SAVINGS_BYTES < original_size:
                    os.replace(temp_path, path)
                    return original_size, optimized_size

                temp_path.unlink(missing_ok=True)
                return original_size, original_size
            except Exception:
                temp_path.unlink(missing_ok=True)
                raise
    except Exception as exc:  # continue optimizing the rest of the deploy package
        print(f"[optimize-images] skipped {path}: {exc}")
        return original_size, original_size


def main() -> None:
    if not PUBLIC_FOTOS_DIR.exists():
        print(f"[optimize-images] {PUBLIC_FOTOS_DIR} not found; nothing to optimize.")
        return

    images = [
        path
        for path in PUBLIC_FOTOS_DIR.rglob("*")
        if path.is_file() and path.suffix.lower() in SUPPORTED_EXTENSIONS
    ]

    before_total = 0
    after_total = 0
    changed = 0

    for image_path in images:
        before, after = optimize_file(image_path)
        before_total += before
        after_total += after
        if after < before:
            changed += 1

    saved = before_total - after_total
    print(
        "[optimize-images] "
        f"processed={len(images)} changed={changed} "
        f"before={before_total / 1048576:.2f}MB "
        f"after={after_total / 1048576:.2f}MB "
        f"saved={saved / 1048576:.2f}MB"
    )


if __name__ == "__main__":
    main()
